jobrunner/lib/database.py

```diff
@@ -150,14 +156,18 @@ def select_values(itemclass, column, **query_params):
     return [decode_field_values(fields, row)[0] for row in cursor]
 
 
-def transaction():
+def transaction(trace=True):
     # Connections function as context managers which create transactions.
     # See: https://docs.python.org/3/library/sqlite3.html#using-the-connection-as-a-context-manager
     # We're relying here on the fact that because of the lru_cache,
     # `get_connection` actually returns the same connection instance every time
-    conn = get_connection()
-    conn.execute("BEGIN IMMEDIATE")
-    return conn
+    span_context = (
+        tracer.start_as_current_span("TRANSACTION") if trace else nullcontext()
+    )
+    with span_context:
```

Comment:
I'm not sure this is going to work as is because the existing code doesn't define a context manager: it relies on the fact that the connection object [acts as a context manager](https://docs.python.org/3/library/sqlite3.html#how-to-use-the-connection-context-manager). So the trace code will make the transaction as finished as soon as this function returns, which is incorrect.

I think we might need to define an explicit context manager to make this work, unless there's another trick I'm missing.

Reply:
Hmm, that is a good point.

I started with a @contextmanager, and then removed it, but yeah this is clearly not going to measure the right thing, good catch.